src/Compilers/Core/Portable/DiagnosticAnalyzer/ArtifactProducer.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.Text;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// The base type for artifact producers that can programmatically generate additional non-code files during a
+    /// compilation.  Artifact generators only run when a compiler is invoked with the <c>generatedfilesout</c>
+    /// parameter.
+    /// </summary>
+    /// <remarks>
+    /// Normally, an artifact generator will not need to report diagnostics.  However, it may sometimes be necessary if
+    /// errors or other issues arise during generation.  If diagnostic reporting is needed, then the same mechanism are
+    /// available as with normal analyzers.  Specifically, <see cref="SupportedDiagnostics"/> should be overridden to
+    /// state which diagnostics may be produced, and the various context `ReportDiagnostic` calls should be used to
+    /// report them.
+    /// </remarks>
+    public abstract class ArtifactProducer : DiagnosticAnalyzer
+    {
+#if !NETCOREAPP
+        /// <summary>
+        /// From: https://github.com/microsoft/referencesource/blob/f461f1986ca4027720656a0c77bede9963e20b7e/mscorlib/system/io/streamwriter.cs#L48
+        /// </summary>
+        private const int DefaultBufferSize = 1024;
+#endif
+
+        /// <summary>
+        /// Callback the compiler can pass into us to actually generate artifacts.  This is safe to hold as a mutable
+        /// value as this is only set once during batch compile as that's the only scenario where /generatedartifactsout
+        /// can be provided.
+        /// </summary>
+        internal Action<string, Action<Stream>>? CreateArtifactStream;
+
+        // By default artifact generators don't report diagnostics.  However, they are still allowed to if they run into
+        // any issues.
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray<DiagnosticDescriptor>.Empty;
+
+        /// <summary>
+        /// Writes out an artifact with the contents of <paramref name="source"/>.  The artifact will be written out with utf8 encoding.
+        /// </summary>
+        /// <param name="fileName">The file name to generate this artifact into.  Will be concatenated with the
+        /// generatedartifactsout path provided to the compiler.</param>
+        /// <param name="source">The text to generate</param>
+        protected void WriteArtifact(string fileName, string source)
+        {
+            WriteArtifact(fileName, stream =>
+            {
+#if NETCOREAPP
```

Comment:
Consider extracting the writer creation, so we don't have the same #if def's three times repeated

Reply:
done.  much nicer!